Fix incorrect choice of accidentals on notes with same midi pitches on capx file import.

Jojo-Schmitz opened the pull request:
Port of #6032, plus adjusting 2 unit tests (both for test 4), which actually prove the fix right.

Resolves: https://musescore.org/en/node/304625 issue b), issue a) seems fixed in MusicXML import since long, but still exists in Capella import

Jojo-Schmitz commented:
No idea why tes78 fails:
```
--- /home/runner/work/MuseScore/MuseScore/src/importexport/capella/tests/data/test7.cap-ref.mscx	2023-12-02 40:54.643392665 +0000
+++ test7.mscx	2023-12-02 14:00:53.023549224 +0000
@@ -417,8 +417,11 @@
           <Chord>
             <durationType>eighth</durationType>
             <Note>
+              <Accidental>
+                <subtype>accidentalFlat</subtype>
+                </Accidental>
               <pitch>61</pitch>
-              <tpc>21</tpc>
+              <tpc>9</tpc>
               </Note>
             </Chord>
           </voice>
   <diff -u /home/runner/work/MuseScore/MuseScore/src/importexport/capella/tests/data/test7.cap-ref.mscx test7.mscx failed, code: 1 
../src/importexport/capella/tests/capella_tests.cpp:64: Failure
```

Jojo-Schmitz commented:
Would it be OK to adjust the test? Or is there s bug in Score::spell(), esp. with that key signature, C#-Major/Ab-minor?

Jojo-Schmitz commented:
So the question is whether there is a bug in`Score::spell()`(or rather in one of the methods it calls), or whether to outsmart it by adjusting the test 7 (test 4 apparently did the same thing prior to this PR ;-))

Jojo-Schmitz committed:
Outsmart `Score::spell()`

Jojo-Schmitz commented:
Fixing a bug in `Score::spell()` would be out of score for this PR, so let's outsmart it.

Jojo-Schmitz commented:
If/when this qualifies for 4.2.0, let me know

Jojo-Schmitz commented:
Any news on this?

Jojo-Schmitz commented:
@cbjeukendrup any news here?

cbjeukendrup commented:
I'm quite uncertain about this. There are two questions:
- Isn't calling `spell` a bit overkill here? It seems a simple per-note contextless algorithm might suffice. But I don't have any experience with Capella at all, so I can't really tell. 
- About that bug in that test: is this really a bug in `spell`, or is it because `spell` is trying to be context-aware while that is not desired in this case?

Jojo-Schmitz commented:
Hmm, to be honest I have no idea.

cbjeukendrup commented:
I'm afraid we'll really have to look for a different solution. Here's a slightly modified version of the test file:
[test01 kopie.capx.zip](https://github.com/musescore/MuseScore/files/14316711/test01.kopie.capx.zip)
<img width="411" alt="Scherm­afbeelding 2024-02-17 om 00 05 06" src="https://github.com/musescore/MuseScore/assets/48658420/111c3c07-ed20-4b53-aa82-6c3eeb6db631">
(that's what it looks like in Capella Reader 8.)
Note the two different enharmonic spellings of the D#/Eb. But both master and your PR import both notes as the same note; master goes for D# and your PR goes for Eb.
Conclusion: this PR is shifting the problem, not solving it, so to speak :)

Jojo-Schmitz commented:
Should we then close this?

Jojo-Schmitz commented:
Hmm, master does this with that capella file:
![image](https://github.com/musescore/MuseScore/assets/1786669/cfe10f60-9e22-4b52-9826-7081ffad603e)

Jojo-Schmitz commented:
so my PR gets 3 right, 1 wrong, master gets 2 right and 2 wrong, to me that is an improvement ;-)

Jojo-Schmitz commented:
Probably this is sort of the same issue as that with test7.cap, 'fixed' in that 2nd commit

Jojo-Schmitz commented on src/importexport/capella/tests/data/test4.capx-ref.mscx:
```diff
@@ -195,10 +195,11 @@
               <eid>m_m</eid>
               <Accidental>
                 <subtype>accidentalFlat</subtype>
+                <subtype>accidentalSharp</subtype>
                 <eid>n_n</eid>
                 </Accidental>
```
oops, an error when resolving merge conflics

Jojo-Schmitz replied:
yep, that's the plan. Will just wait for the unit tests to finish, to see whether there's more to fix

Jojo-Schmitz commented on src/importexport/capella/internal/capella.cpp:
```diff
@@ -979,6 +978,7 @@ static Fraction readCapVoice(Score* score, CapVoice* cvoice, int staffIdx, const
         }
     }
     Fraction endTick = tick;
+    score->spell(); // Call respell-pitches to correct accidentals
```
Shouldn't it then be even outside the `staves` loop (starting at line1360)?

Jojo-Schmitz committed:
Optimization, call `spell()` only once per score, not per system, staff and voice

Jojo-Schmitz committed:
[MU4] Fix incorrect choice of accidentals on notes with same midi pitches on capx file import.

Port of #6032, plus adjusting 2 unit tests (both for test 4), which actually prove the fix right.
